packages/ember-glimmer/lib/ember-template-compiler/plugins/transform-action-syntax.js

```diff
@@ -39,17 +39,29 @@ TransformActionSyntax.prototype.transform = function TransformActionSyntax_trans
 
   traverse(ast, {
     ElementModifierStatement(node) {
-      if (TRANSFORMATIONS[node.path.original]) {
-        let thisPath = b.path('this');
-
-        // We have to delete the `parts` here because otherwise it will be treated
-        // as a property look up (i.e. `this.this`) and will result in `undefined`.
-        thisPath.parts = [];
-
-        return b.elementModifier(node.path, [thisPath, ...node.params], node.hash, node.loc);
+      if (isAction(node)) {
+        insertThisAsFirstParam(node, b);
+      }
+    },
+    MustacheStatement(node) {
+      if (isAction(node)) {
+        insertThisAsFirstParam(node, b);
+      }
+    },
+    SubExpression(node) {
+      if (isAction(node)) {
+        insertThisAsFirstParam(node, b);
       }
     }
   });
 
   return ast;
 };
+
+function isAction(node) {
+  return (node.type === 'MustacheStatement' || node.type === 'SubExpression' || node.type === 'ElementModifierStatement') && (node.path.original === 'action');
```

Comment:
I don't think you need to check the `node.type` anymore. `traverse` does that for you already.